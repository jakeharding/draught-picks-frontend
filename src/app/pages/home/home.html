<!--
  Generated template for the HomePage page.

  See http://ionicframework.com/docs/components/#navigation for more info on
  Ionic pages and navigation.
-->
<ion-header>
  <draught-header></draught-header>
</ion-header>

<ion-content class="ion-padding">
  <ion-row>
    <ion-col class="below-header" size="12" sizeSm="10" offsetSm="1" sizeMd="8" offsetMd="2" sizeLg="6" offsetLg="3" sizeXl="6" offsetXl="3">

      <ion-segment color="primary" [(ngModel)]="isRecentBeersSelected">
        <ion-segment-button tid="recentSegment" value="yes">
          Recent
        </ion-segment-button>
        <ion-segment-button tid="recommendedSegment" checked value="no">
          Recommended
        </ion-segment-button>
      </ion-segment>
      <label class="beerError" *ngIf="BACcalc">You have had {{totalCount}} beers in the past 3 hours. You might want to find a ride home.</label>
      <div [ngSwitch]="isRecentBeersSelected">

        <div *ngSwitchCase="'yes'">
          <div class="ion-text-center" *ngIf="recents?.length === 0">
            <p>You haven't had any beers yet! <a tid="toSearchPage" class="pointer" routerLink="/search">Search for what you are drinking.</a></p>
          </div>
          <div *ngIf="recents?.length > 0">

            <strong class="title">Your recent beers:</strong>
            <hr>
<!--            TODO use ionic 4 infinte scroll -->
            <div class="list-container" id="infinite-scroller1" appInfiniteScroller [immediateCallback]="false" [scrollCallback]="recentScrollCallback" [loadMore]="loadMoreRecent">
              <beer-list [beers]="recents"></beer-list>
            </div>
          </div>
        </div>
        <div *ngSwitchCase="'no'">
            <div class="ion-text-center" *ngIf="recommended?.length === 0">
              <p>Your recommended beers are coming soon! If you haven't created your beer profile yet,
                <a class="pointer" routerLink="/preferences">click here to fill it out now!</a></p>
            </div>
            <div *ngIf="recommended?.length > 0">
              <strong class="title">Your recommended beers:</strong>
              <hr>
              <div class="list-container" id="infinite-scroller2" appInfiniteScroller [immediateCallback]="false" [scrollCallback]="recScrollCallback" [loadMore]="loadMoreRecommended">
                <beer-list [beers]="recommended"></beer-list>
              </div>
            </div>
        </div>
      </div>

    </ion-col>
  </ion-row>
</ion-content>
