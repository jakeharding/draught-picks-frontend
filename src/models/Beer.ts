/**
 * Beer.ts
 *
 * Created by jake
 * Created on 2/27/18
 *
 * Model the beers
 */
import BeerRating from "./BeerRating";


export default interface Beer {
  name: string;
  description: string,
  abv: number;
  ibu: number;
  uuid: string;
  api_id: string;
  name_of_api: string;
  created_at: Date;
  rating?: Array<BeerRating>;
}
