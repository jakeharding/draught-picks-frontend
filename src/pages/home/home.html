<!--
  Generated template for the HomePage page.

  See http://ionicframework.com/docs/components/#navigation for more info on
  Ionic pages and navigation.
-->
<ion-header>
  <ion-navbar color="primary">
    <ion-title text-center>Draught Picks</ion-title>
  </ion-navbar>
</ion-header>

<ion-content padding>
  <ion-row>
    <ion-col col-12 col-sm-10 offset-sm-1 col-md-8 offset-md-2 col-lg-6 offset-lg-3 col-xl-6 offset-xl-3>
      <div text-center *ngIf="recents?.length === 0">
        <p>You haven't had any beers yet! <a class="pointer"(click)="navCtrl.parent.select(0)">Search for what you are drinking.</a>
        </p>
      </div>

      <div *ngIf="recents?.length > 0">
        <strong class="title">Your recent beers:</strong>
        <hr>
        <beer-list [beers]="recents"></beer-list>
      </div>
    </ion-col>
  </ion-row>
</ion-content>
