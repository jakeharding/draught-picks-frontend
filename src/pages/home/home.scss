page-home {
  .title {
    font-size: 2em;
  }
  ion-icon{
    font-family: sans-serif;
  }
  a{
    text-decoration: none;
  }
  .list-container {
    overflow-y: auto;
    position: relative;
    height: 600px;
  }
}
