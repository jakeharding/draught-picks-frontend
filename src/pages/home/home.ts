import { Component } from '@angular/core';
import { IonicPage, NavController, NavParams } from 'ionic-angular';
import {LoginRequired} from "../../providers/auth/auth";
import {PreferencesProvider} from "../../providers/preferences/preferences";
import UserPreferences from "../../models/UserPreferences";

/**
 * Generated class for the HomePage page.
 *
 * See https://ionicframework.com/docs/components/#navigation for more info on
 * Ionic pages and navigation.
 */

@LoginRequired
@IonicPage()
@Component({
  selector: 'page-home',
  templateUrl: 'home.html',
})
export class HomePage {
  prefs: UserPreferences;

  constructor(public navCtrl: NavController, public navParams: NavParams,
              public prefsProvider: PreferencesProvider) {
    this.prefsProvider.retrieve().then((prefs: UserPreferences) => {
      this.prefs = prefs;
    });
  }
}
