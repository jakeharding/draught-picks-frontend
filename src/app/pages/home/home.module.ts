import { NgModule } from '@angular/core';
import { IonicModule } from '@ionic/angular';
import { HomePage } from './home';
import { ComponentsModule } from '../../components/components.module';
import { RouterModule } from '@angular/router';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';

@NgModule({
  declarations: [
    HomePage,
  ],
  imports: [
    IonicModule,
    RouterModule.forChild([{path: 'home', component: HomePage}]),
    ComponentsModule,
    CommonModule,
    FormsModule
  ],
})
export class HomePageModule {}
