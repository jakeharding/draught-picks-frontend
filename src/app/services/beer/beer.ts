import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { environment as Env } from '../../../environments/environment';
import Beer from '../../models/Beer';
import PageResponse from '../../models/PageResponse';
import { LIMIT } from '../../directives/infinite-scroller/infinite-scroller';
import { EMPTY, Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { AutoCompleteService } from 'ionic4-auto-complete';
/*
  Generated class for the BeerProvider provider.

  See https://angular.io/guide/dependency-injection for more info on providers
  and Angular DI.
*/
@Injectable({
  providedIn: 'root',
})
export class BeerProvider implements AutoCompleteService {
  url: string;
  recentsUrl: string;
  recommendedUrl: string;
  labelAttribute = 'name'; // Attribute needed for the AutoCompleteService
  constructor(public http: HttpClient) {
    this.url = `${Env.REST_API_ROOT}/beers`;
    this.recentsUrl = `${Env.REST_API_ROOT}/recent-beers`;
    this.recommendedUrl = `${Env.REST_API_ROOT}/recommended-beers`;
  }

  /**
   * search function
   * Parameters: Observable of type Beer list
   * Sets the limit and offset for the infinite scrolling
   */
  search(params: any): Observable<Beer[]> {
    if (!params.limit) {
      params.limit = LIMIT;
    }
    if (!params.offset) {
      params.offset = 0;
    }
    return this.http.get(this.url, {params}).pipe(map(({results}: PageResponse<Beer>) => results));
  }

  /**
   * Method called in auto complete search bar.
   * @param beerName - Name of beer
   * @returns - List of beers
   */
  getResults(beerName: string) {
    if (beerName) {
      return this.search({search: beerName});
    }
    return EMPTY;
  }

  /**
   * recents function
   * Parameters: Observable of type Beer list
   * Sets the limit and offset for the infinite scrolling
   */
  recents(params: any): Observable<Beer[]> {
    if (!params.limit) {
      params.limit = LIMIT;
    }
    if (!params.offset) {
      params.offset = 0;
    }

    return this.http.get(this.recentsUrl, {params}).pipe(map(({results}: PageResponse<Beer>) => results));
  }

  /**
   * recommended function
   * Parameters: query params for pagination
   * Retrieves a page of recommended beers for an authenticated user.
   */
  recommended(params: any): Observable<Beer[]> {
    if (!params.limit) {
      params.limit = LIMIT;
    }
    if (!params.offset) {
      params.offset = 0;
    }
    return this.http.get(this.recommendedUrl, {params})
      .pipe(map(({results}: PageResponse<Beer>) => results));
  }

  /**
   * createRecent function
   * Parameters: Beer
   * returns the recent beer list
   */
  createRecent(recent: Beer) {
    return this.http.post(this.recentsUrl, {beer: recent.uuid}).toPromise();
  }

  /**
   * Retrieve a information about a beer.
   * @param uuid - ID of beer
   * @returns - Promise a beer
   */
  retrieve(uuid: string) {
    return this.http.get(`${this.url}/${uuid}`).toPromise();
  }


}
