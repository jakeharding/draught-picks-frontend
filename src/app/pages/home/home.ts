import { Component } from '@angular/core';
import { NavController, NavParams } from '@ionic/angular';
import Beer from '../../models/Beer';
import { BeerProvider } from '../../services/beer/beer';
import { EMPTY } from 'rxjs';
import { BasePage } from '../BasePage';
import { LIMIT } from '../../app.component';


/**
 * Generated class for the HomePage page.
 *
 * See https://ionicframework.com/docs/components/#navigation for more info on
 * Ionic pages and navigation.
 */

@Component({
  selector: 'page-home',
  templateUrl: 'home.html',
  styleUrls: ['./home.scss']
})
export class HomePage extends BasePage {
  recents: Array<Beer>;
  recommended: Array<Beer>;
  beerForABV: Array<Beer>;
  BACcalc: boolean;
  totalCount: number;
  isRecentBeersSelected = 'no';
  recommendedOffset: number;
  recentOffset: number;
  loadMoreRecommended: boolean;
  loadMoreRecent: boolean;
  recScrollCallback: () => void;
  recentScrollCallback: () => void;


  /**
   * HomePage constructor
   * Required constructor for the HomePage class, sets the offset, loadMoreRecommended, loadMoreRecent,
   * and recommended values
   */
  constructor(public navCtrl: NavController,
              public beerProvider: BeerProvider) {
    super('home');
    this.recommendedOffset = 0;
    this.recentOffset = 0;
    this.loadMoreRecommended = true;
    this.loadMoreRecent = true;
    this.recommended = [];
    this.recents = [];
  }

  /**
   * ionViewWillEnter function
   * No Parameters
   * Gets the recent and recommended beers and sets the LIMIT and offset values for the scrolling functionality.
   * Also calls the scrolling functionality
   */
  ionViewWillEnter() {
    const time = new Date();
    time.setHours(time.getHours() - 3);
    const pastTime =  new Date(time);

    this.beerProvider.recents({limit: LIMIT, offset: 0}).toPromise().then(results => {

      this.recents = results;
      this.beerForABV = this.recents.filter(beer => {
        beer.recents = beer.recents.filter( currentBeer => {

          return (new Date(currentBeer.created_at)) > pastTime;
        });
        return beer.recents.length > 0;
      });

      this.totalCount = 0;
      if ( this.beerForABV.length > 0) {
        this.totalCount = this.beerForABV.map((current) => current.recents.length).reduce((acc, c) => {
          return acc + c;
        });
      }
      this.BACcalc = this.totalCount >= 4;
    });

    this.beerProvider.recommended({limit: LIMIT, offset: 0}).toPromise().then(results => {
      this.recommended = results;
    });

  }

  /**
   * Get more recent beers if available.
   * @returns - recents
   */
  getRecentBeers(event) {
    if (this.loadMoreRecent) {
      const queryParams = {
        limit: LIMIT,
        offset: this.recentOffset
      };
      return this.beerProvider.recents(queryParams).toPromise().then(beers => this.processRecentBeers(beers, event));
    }
    return EMPTY;
  }

  /**
   * Get more recommended beers if available.
   * @returns - recommended
   */
  getRecommendedBeers(event) {
    if (this.loadMoreRecommended) {
      const queryParams = {
        limit: LIMIT,
        offset: this.recommendedOffset
      };
      return this.beerProvider.recommended(queryParams).toPromise().then(beers => this.processRecommendedBeers(beers, event));
    }
    return EMPTY;

  }

  /**
   * processRecommendedBeers function
   * Parameters: Beer list
   * increases the offset and loads more beers while scrolling through the beer list
   */
  private processRecommendedBeers = (beers, event) => {
    event.target.complete();
    if (beers.length === 0) {
      this.loadMoreRecommended = false;
      return;
    }
    this.recommendedOffset += LIMIT;
    this.recommended = this.recommended.concat(beers);
  }


  /**
   * processRecentBeers function
   * Parameters: Beer list
   * increases the offset and loads more beers while scrolling through the beer list
   */
  private processRecentBeers = (beers, event) => {
    event.target.complete();
    if (beers.length === 0) {
      this.loadMoreRecent = false;
      return;
    }
    this.recentOffset += LIMIT;
    this.recents = this.recents.concat(beers);
  }

}
