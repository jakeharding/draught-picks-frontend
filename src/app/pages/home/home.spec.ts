import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { HomePage } from './home';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { NavController, NavParams } from '@ionic/angular';
import { InfiniteScrollerDirective, LIMIT } from '../../directives/infinite-scroller/infinite-scroller';
import { BeerProvider } from '../../services/beer/beer';
import { Observable } from 'rxjs';
import Beer from '../../models/Beer';
import RecentBeer from '../../models/RecentBeers';

/**
 * home.spec.ts
 *
 * Created by jake
 * Created on 11/18/18
 *
 * Test HomePage.
 */

describe('HomePage', () => {
  let fixture: ComponentFixture<HomePage>;
  let component: HomePage;

  const mockRecentBeer = {created_at: new Date(), uuid: 'recent', user: 'user', beer: 'beer'} as RecentBeer;
  const mockBeer = { uuid: 'beer', recents: [mockRecentBeer]} as Beer;

  const mockBeerProvider = {
    recents: jest.fn(() => Observable.of([mockBeer])),
    recommended: jest.fn(() => Observable.of([{} as Beer]))
  };
  beforeEach(async(() => {
    TestBed.configureTestingModule({
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
      declarations: [HomePage, InfiniteScrollerDirective],
      providers: [
        { provide: NavController, useValue: {}},
        { provide: NavParams, useValue: {}},
        { provide: BeerProvider, useValue: mockBeerProvider}
      ]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(HomePage);
    component = fixture.debugElement.componentInstance;
  });

  it('should initialize the HomePage', () => {
    expect(component).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });

  test('properties are set after initialization', () => {
    expect(component.loadMoreRecommended).toEqual(true);
    expect(component.loadMoreRecent).toEqual(true);
    expect(component.recScrollCallback).toBeDefined();
    expect(component.recentScrollCallback).toBeDefined();
  });

  test('ionViewWillEnter sets up the BAC calculation', async () => {
    // BAC calculation may be removed so only adding lightweight testing to get coverage
    await component.ionViewWillEnter();
    expect(mockBeerProvider.recents).toHaveBeenCalledTimes(1);
    expect(mockBeerProvider.recommended).toHaveBeenCalledTimes(1);
  });

  test('getRecentBeers calls to get more recent beers', () => {
    component.loadMoreRecent = true;
    component.getRecentBeers();
    expect(mockBeerProvider.recents).toHaveBeenCalledTimes(1);
  });

  test('getRecentBeers returns empty Observable', () => {
    component.loadMoreRecent = false;
    const result = component.getRecentBeers();
    expect(mockBeerProvider.recents).not.toHaveBeenCalled();
    expect(result).toEqual(Observable.empty());
  });

  test('getRecommendedBeers calls to get more recommended beers', () => {
    component.loadMoreRecommended = true;
    component.getRecommendedBeers();
    expect(mockBeerProvider.recommended).toHaveBeenCalledTimes(1);
  });

  test('getRecommendedBeers returns empty Observable', () => {
    component.loadMoreRecommended = false;
    const result = component.getRecommendedBeers();
    expect(mockBeerProvider.recommended).not.toHaveBeenCalled();
    expect(result).toEqual(Observable.empty());
  });

  test('processRecommendedBeers returns undefined', () => {
    const result = component.processRecommendedBeers([]);
    expect(result).toBeUndefined();
    expect(component.loadMoreRecommended).toBe(false);
  });

  test('processRecommendedBeers increases offset and adds to the recommended', () => {
    const mockBeers = [{} as Beer];
    const offset = component.recommendedOffset;
    component.processRecommendedBeers(mockBeers);
    expect(component.recommendedOffset).toBe(offset + LIMIT);
    expect(component.recommended).toContain(mockBeers[0]);
  });

  test('processRecentBeers returns undefined', () => {
    const result = component.processRecentBeers([]);
    expect(result).toBeUndefined();
    expect(component.loadMoreRecent).toBe(false);
  });

  test('processRecentBeers increases offset and adds to the recents', () => {
    const mockBeers = [{} as Beer];
    const offset = component.recommendedOffset;
    component.processRecentBeers(mockBeers);
    expect(component.recentOffset).toBe(offset + LIMIT);
    expect(component.recents).toContain(mockBeers[0]);
  });
});
