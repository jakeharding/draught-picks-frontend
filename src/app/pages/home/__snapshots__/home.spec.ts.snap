// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage should initialize the HomePage 1`] = `
<page-home
  beerProvider={[Function Object]}
  isRecentBeersSelected={[Function String]}
  loadMoreRecent={[Function Boolean]}
  loadMoreRecommended={[Function Boolean]}
  navCtrl={[Function Object]}
  processRecentBeers={[Function Function]}
  processRecommendedBeers={[Function Function]}
  recentOffset="0"
  recents={[Function Array]}
  recommended={[Function Array]}
  recommendedOffset="0"
>
  <ion-header>
    <draught-header />
  </ion-header><ion-content
    class="ion-padding"
  >
    <ion-row>
      <ion-col
        class="below-header"
        offsetlg="3"
        offsetmd="2"
        offsetsm="1"
        offsetxl="3"
        size="12"
        sizelg="6"
        sizemd="8"
        sizesm="10"
        sizexl="6"
      >
        <ion-segment
          color="primary"
        >
          <ion-segment-button
            tid="recentSegment"
            value="yes"
          >
             Recent 
          </ion-segment-button>
          <ion-segment-button
            checked=""
            tid="recommendedSegment"
            value="no"
          >
             Recommended 
          </ion-segment-button>
        </ion-segment>
        
        <div>
          
          
        </div>
      </ion-col>
    </ion-row>
  </ion-content>
</page-home>
`;
